### PR TITLE
frontend: return empty dict instead of null in API calls to BE

### PR DESCRIPTION
**Pull request by nikromen**
Fixes: #2816

### CODE REVIEW

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py [hunk 1]
```diff
@@ -354,6 +354,9 @@ def get_build_task(task_id):
     try:
         task = BuildsLogic.get_build_task(task_id)
         build_record = get_build_record(task)
+        if build_record is None:
```
Comment (nikromen) [inline frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:357]: I can see that `get_build_record` and `get_srpm_build_record` are only used in API endpoint. Can we perhaps change return type of these to `dict` from `Optional[dict]`? Backend thinks these endpoints won't return `null` but they do

Comment (praiskup) [reply frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:360]: I can't reproduce the `null` output, I can't decide.  It seems completely OK to return 404 if no such build task exists, 500 upon error, 400 for invalid requests.  And return 200 if and only if the request was processed.

**Comment by praiskup** [comment]
I'm for fixing the `Backend <-> Frontend` protocol, though this rather seems like a `status 200` by mistake, instead of `status 404`.  These additional checks seem pretty complicated to validate.

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py (hunk 1 above)
Comment (nikromen) [reply frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:360]: the build that causes this error is https://copr.fedorainfracloud.org/coprs/rhcontainerbot/podman-next/build/5942294/

namely, the f36-s390x chroot, as you can see there the chroot for this build does not exist... it looks like it was never created in DB for this build - I can see that this build was created at the same time I made f36 chroots EOL so some rare unlucky race condition perhaps :D

Until that time the build is still in a pending state and it is flooding our logs with errors (cca 2 months).

**Comment by nikromen** [comment]
I fix the `null` output on the frontend API endpoints side - but the race condition with EOL and new builds remains... it seems like a really rare accident so I think it's not worth spending time on it, let's see if it will occur in next EOL session

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py (hunk 1 above)
Comment (praiskup) [reply frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:357]: Hmmm, that would make sense, good catch.

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py [hunk 2]
```diff
@@ -354,6 +354,14 @@ def get_build_task(task_id):
...
+            jsonout = flask.jsonify(
+                {"msg": "Internal server error happened during searching"
+                        f" record for task {task_id} was not found"}
+            )
+            jsonout.status_code = 500
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:362]: Can we use error 400?  And I'm still not quite sure this is ideal, can't we raise this error directly in `get_build_task()`?

Comment (nikromen) [reply frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:362]: OK, I used the already existing `get_copr_chroot_safe` method which returns 404, and changed the `except: return None` to re-raise - hopefully backend can handle this :D

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py [hunk 3]
```diff
@@ -171,7 +170,7 @@ def get_build_record(task, for_backend=False):
             "repo_priority": task.build.copr.repo_priority
         })
 
-        copr_chroot = CoprChrootsLogic.get_by_name_safe(task.build.copr, task.mock_chroot.name)
+        copr_chroot = ComplexLogic.get_copr_chroot_safe(task.build.copr, task.mock_chroot.name)
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:173]: This deserves a note in the commit message. :-)

**Comment by praiskup** [comment]
Hm, while it looks like it should work, it doesn't seem to.

I just tested locally, and I'm still getting:
```
ERROR:coprs:'NoneType' object has no attribute 'module_setup_commands'
Traceback (most recent call last):
  File "/copr/coprs_frontend/coprs/views/backend_ns/backend_general.py", line 175, in get_build_record
    modules = copr_chroot.module_setup_commands
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'module_setup_commands'
```

And backend gets:
```
$ curl localhost:55555/backend/get-build-task/95628-rhel-9-x86_64
null
```

**Comment by nikromen** [comment]
I did an eol event during SRPM build (to simulate https://github.com/fedora-copr/copr/pull/2842#discussion_r1291348851) to test the `Backend <-> Frontend` protocol and here's what I got:

FE:
```
2023-09-19 15:26:40,070 [INFO][/opt/copr/frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:357|backend_general:_stream][backend: 10.89.0.13] Generating RPM builds
2023-09-19 15:26:40,502 [ERROR][/opt/copr/frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:192|backend_general:get_build_record][backend: 10.89.0.13] Chroot name fedora-38-x86_64 does not exist in Copr nikromen/dalsi-novy.
Traceback (most recent call last):
  File "/opt/copr/frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py", line 173, in get_build_record
    copr_chroot = ComplexLogic.get_copr_chroot_safe(task.build.copr, task.mock_chroot.name)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/copr/frontend/coprs_frontend/coprs/logic/complex_logic.py", line 289, in get_copr_chroot_safe
    raise ObjectNotFound(
coprs.exceptions.ObjectNotFound: Chroot name fedora-38-x86_64 does not exist in Copr nikromen/dalsi-novy.
```

BE:
```
[2023-09-19 15:31:40,518][ ERROR][PID:353][/usr/bin/copr-backend-process-build.managed.pid-353][background_worker_build.py:_get_build_job:342] Failed to download build info: Request client error on http://frontend:5000/backend/get-build-task/27-fedora-38-x86_64/: 404 NOT FOUND
[2023-09-19 15:31:40,518][ ERROR][PID:353][/usr/bin/copr-backend-process-build.managed.pid-353][background_worker_build.py:handle_task:793] Backend process error: Failed to get the build task get-build-task/27-fedora-38-x86_64
[2023-09-19 15:31:40,519][ ERROR][PID:353][/usr/bin/copr-backend-process-build.managed.pid-353][background_worker_build.py:handle_task:804] No job object from Frontend
```

so now as expected, frontend returns 404 and backend can handle it thus logs look meaningful now. But yep, the race condition with eol, etc. events still remain (I would rather do a separate issue/PR about it)

File: frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py [hunk 4]
```diff
@@ -191,7 +190,7 @@ def get_build_record(task, for_backend=False):
 
     except Exception as err:
         app.logger.exception(err)
-        return None
+        raise err
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/views/backend_ns/backend_general.py:193]: This is weird.  app.logger.exception dumps the traceback into logs for the first time, and if we re-raise - some other exception handler will log it once more?

**Review by praiskup — APPROVED** [review]
Seems reasonable to me. Thank you!

**Pull request merged**